packages/monitoring-system/data-processor/test/unit/data-processors/cloud-logs-data-processor.spec.ts

```diff
@@ -0,0 +1,601 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+import {describe, it, beforeEach} from 'mocha';
+import assert from 'assert';
+import {MockSubscription} from './mocks/mock-pubsub-subscription';
+import {PubSub} from '@google-cloud/pubsub';
+import {MockFirestore, MockRecord} from './mocks/mock-firestore';
+import {CloudLogsProcessor} from '../../../src/data-processors/cloud-logs-data-processor';
+import {loadFixture} from './util/test-util';
+import {ObjectWritableMock} from 'stream-mock';
+import pino from 'pino';
+
+let mockSubscription: MockSubscription;
+let mockFirestore: MockFirestore;
+let processor: CloudLogsProcessor;
+
+const VALID_MESSAGES: {[name: string]: {}} = loadFixture(
+  'mock-pubsub-log-messages.json',
+  false
+);
+
+const MALFORMED_MESSAGES: {[name: string]: {}} = loadFixture(
+  'mock-pubsub-log-messages-malformed.json',
+  false
+);
+
+/**
+ * Returns the given object as a Buffer
+ * @param obj object to convert
+ */
+function asBuffer(obj: {}): Buffer {
+  return Buffer.from(JSON.stringify(obj));
+}
+
+/**
+ * Returns a logger with the given destination
+ * @param dest an ObjectWritableMock stream
+ */
+function getMockLogger(dest: ObjectWritableMock): pino.Logger {
+  const defaultOptions: pino.LoggerOptions = {
+    base: null,
+    messageKey: 'message',
+    timestamp: false,
+    level: 'trace',
+  };
+  return pino(defaultOptions, dest);
+}
+
+/**
+ * Asserts that the given writestream has an error logged
+ * with the given message
+ */
+function assertErrorLogged(
+  expectedErrorMsg: string,
+  writeStream: ObjectWritableMock
+) {
+  try {
+    writeStream.end();
+    const lines: string[] = writeStream.data;
+    const jsonLines = lines.map(line => JSON.parse(line));
+    assert(jsonLines.length > 0, 'No logs found');
+    let foundErrorLog = false;
+    for (const line of jsonLines) {
+      const isErrorLevel = line.level === 50;
+      const hasCorrectMsg =
+        line.message && String(line.message).includes(expectedErrorMsg);
+      foundErrorLog = isErrorLevel && hasCorrectMsg;
+      if (foundErrorLog) break;
+    }
+    assert(foundErrorLog, 'No relevant errors were logged');
+  } catch (error) {
+    throw new Error(`Failed to read stream: ${error}`);
+  }
+}
+
+/**
+ * Asserts that the given message was properly processed and acked.
+ * @param message message to test
+ * @param expectedRecords expected records in firestore
+ * @param preExistingRecords records to add to firestore before test
+ */
+async function testValidMessage(
+  message: {},
+  expectedRecords: MockRecord[],
+  preExistingRecords?: MockRecord[]
+): Promise<void> {
+  if (preExistingRecords) {
+    preExistingRecords.forEach(record => mockFirestore.addRecord(record));
+  }
```

Comment:
I would probably do this in the preamble of the test - setting up the conditions of the test rather than putting it in this test method.